// @react-pdf/core/lib/stylesheet/index.js
'use strict';

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _mediaEngine = require('media-engine');

var _mediaEngine2 = _interopRequireDefault(_mediaEngine);

var _transformStyles = require('./transformStyles');

var _transformStyles2 = _interopRequireDefault(_transformStyles);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var create = function create(styles) {
  return styles;
};

var flatten = function flatten(input) {
  if (!Array.isArray(input)) {
    input = [input];
  }

  var result = input.reduce(function (acc, style) {
    if (style) {
      Object.keys(style).forEach(function (key) {
        if (style[key]) {
          acc[key] = style[key];
        }
      });
    }

    return acc;
  }, {});

  return result;
};

var resolveMediaQueries = function resolveMediaQueries(input, container) {
  var result = Object.keys(input).reduce(function (acc, key) {
    if (/@media/.test(key)) {
      return _extends({}, acc, (0, _mediaEngine2.default)(_defineProperty({}, key, input[key]), container));
    }

    return _extends({}, acc, _defineProperty({}, key, input[key]));
  }, {});

  return result;
};

var resolve = function resolve(styles, container) {
  if (!styles) {
    return null;
  }

  styles = flatten(styles);
  styles = resolveMediaQueries(styles, container);

  return (0, _transformStyles2.default)(styles);
};

var absoluteFillObject = {
  position: 'absolute',
  top: 0,
  left: 0,
  bottom: 0,
  right: 0
};

module.exports = {
  hairlineWidth: 1,
  create: create,
  resolve: resolve,
  flatten: flatten,
  absoluteFillObject: absoluteFillObject
};
